public/app/features/datasources/settings/CustomHeadersSettings.tsx

```diff
@@ -0,0 +1,174 @@
+import React, { PureComponent } from 'react';
+import { Button, FormField, SecretFormField } from '@grafana/ui/src';
+
+enum HeaderChangeEvent {
+  add,
+  update,
+  reset,
+  remove,
+}
+
+export interface CustomHeader {
+  name: string;
+  value: string;
+  configured: boolean;
+}
+
+export type CustomHeaders = CustomHeader[];
+
+export interface Props {
+  datasource: {
+    jsonData: {
+      [key: string]: any;
+    };
+    secureJsonData: {
+      [key: string]: string;
+    };
+    secureJsonFields: {
+      [key: string]: boolean;
+    };
+  };
+  onChange: (headers: CustomHeaders) => void;
+}
+
+export interface State {
+  headers: CustomHeaders;
+}
+
+export class CustomHeadersSettings extends PureComponent<Props, State> {
+  state = {
+    headers: [],
+  };
+
+  constructor(props: Props) {
+    super(props);
+  }
+
+  componentDidMount(): void {
+    this.updateHeaders();
+  }
+
+  onHeaderChange(
```

Comment:
I suggest refactoring this method into 4 separate ones:

- `onHeaderAdd`
- `onHeaderChange`
- `onHeaderReset`
- `onHeaderRemove`

It will improve the readability of this logic. You will be able to remove the `resetHeader` and `updateHeader` functions and add those implementations respectively in `onHeaderReset`, `onHeaderChange`